### PR TITLE
New plugin: splitbutton

### LINKED ISSUE #1679: Split button UI component
## Are you reporting a feature request or a bug?

Feature request

## Provide detailed reproduction steps (if any)

We need a UI component that would allow to group multiple related options in a single button (in that sense it's similar to menu button). Once some option is activated it should change the icon to the active option.

For example think about alignment buttons. You have four buttons, but only one is active at a time. Having 4 buttons consumes a lot of space, instead we could group it into a single button. Important thing is that upon refresh, the active button is displayed.

We did a similar thing on our CKSource XMass hackathon.

An would look like that:

Split button with justify option active

When thinking about that I'm wondering if we really need a new component, or whether we could enhance menubutton in a way that it would conveniently handle this case.

### PR DESCRIPTION
**Pull request by engineering-this**
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

I've added the `splitbutton` plugin which is extending the `menubutton` class. Also I've slightly updated the `button`

Closes #1679.

### CODE REVIEW

File: tests/plugins/splitbutton/splitbutton.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+						icon: 'superscript'
+					} ]
+				} );
+			}
+		}
+	}
+};
+
+bender.test( {
+	'test splitbutton in toolbar': function() {
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply tests/plugins/splitbutton/splitbutton.js]: There are also missing tests for changing the state of menu items.

File: tests/plugins/splitbutton/manual/splitbutton.html [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+			} );
+		}
+	} );
+
+	CKEDITOR.replace( 'editor', {
+		extraPlugins: 'teststylesplit'
+	} );
+
+	// Just hiding buttons that are irrelevant to test.
+	CKEDITOR.instances.editor.on( 'contentDom', function() {
```
[the remark this answers is not in the document]
Comment (engineering-this) [reply tests/plugins/splitbutton/manual/splitbutton.html]: Hej, @mlewand what is your view on this?

**Comment by engineering-this** [comment]
@Comandeer 

>There is issue with plugins with options that are exclusive (e.g. justify) – menu does not correctly update states.

It looks fine with justify on sample
https://ckeditor4-preview.ckeditor.com/t/1679/plugins/splitbutton/samples/splitbutton.html

>The other concern I have is the fact how menu works: it focuses the first element upon opening. However our Styles or Format combos focuses the first selected option. I'm wondering what would be more user-friendly. Speaking in terms of ARIA: menu or listbox?

I think we need a bit of discussion here if first item should be initially focused or not?

@mlewand What is your tough on that?

>It also uncovers very important issue with the whole implementation: as #678 is not resolved, splitbutton needs other buttons to be added to be able to use their commands. That makes the whole feature de facto not usable at all.

Actually it is caused by ACF. I had to set `config.allowedContent = true;` and everything works. My question here is how should we deal with that?

**Comment by engineering-this** [comment]
After some talking first item will be focused on menu open, just like other menu buttons.

File: tests/plugins/splitbutton/manual/splitbutton.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (mlewand) [reply tests/plugins/splitbutton/manual/splitbutton.html]: For sure we should not workaround this like that. We need a proper fix mentioned by @Comandeer.

I mean, workaround can be placed for the time being, but the issue can not be merged to stable release with such workaround in place.

@engineering-this please, put a reference to #678 anywhere where such workaround is used.

File: plugins/splitbutton/plugin.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+				if ( activeButton ) {
+					CKEDITOR.document.getById( properties.buttons[ defaultButton.id ]._.id ).setStyle( 'display', 'none' );
+					properties.buttons[ defaultButton.id ].hidden = true;
+
+					CKEDITOR.document.getById( properties.buttons[ activeButton.id ]._.id ).removeStyle( 'display' );
+					properties.buttons[ activeButton.id ].hidden = false;
+					previousButton = activeButton;
+				} else {
+					CKEDITOR.document.getById( properties.buttons[ defaultButton.id ]._.id ).removeStyle( 'display' );
```
Comment (Comandeer) [inline plugins/splitbutton/plugin.js]: This code is the same as code above and similar to code, in which `display` is set to `none`. It'd move it to separate, dedicated function.

File: plugins/splitbutton/plugin.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+			// First button as default. It might be overwritten by actual default button.
+			if ( !defaultButton ) {
+				previousButton = properties.buttons[ item.id ];
+				defaultButton = properties.buttons[ item.id ];
+			} else {
+				if ( item[ 'default' ] ) {
+					defaultButton = properties.buttons[ item.id ];
+				} else {
+					properties.buttons[ item.id ].style = 'display: none;';
+					properties.buttons[ item.id ].hidden = true;
```
Comment (Comandeer) [inline plugins/splitbutton/plugin.js]: I understand that it's a workaround for #678 to add features by adding buttons and hiding them?

Comment (engineering-this) [reply plugins/splitbutton/plugin.js]: Correct.

**Comment by engineering-this** [comment]
>Actually we haven't defined any behavior for that, but changing icons in such way seems to be confusing. Microsoft Word has always the same icon in split button face and it changes only its state, e.g. "Numbering" has always the same icon despite which numbering option is selected. The face is used for switching feature on/off, not for representing currently selected option.

This was the our idea of split button, to create 'splitbutton' with static face we could just add one button for default behaviour and next to it menu button for options.

**Comment by Comandeer** [comment]
But now the face is not static, it's dynamically changing and I think it's quite confusing to the end user.

**Comment by engineering-this** [comment]
Well, I think we could add option to lock face on default button.

File: CHANGES.md [hunk 5]
```diff
@@ -6,6 +6,7 @@
 New Features:
 
 * [#1761](https://github.com/ckeditor/ckeditor-dev/issues/1761): [Autolink](https://ckeditor.com/cke4/addon/autolink) plugin supports email links.
+* [#1679](https://github.com/ckeditor/ckeditor-dev/issues/1679): Introduced the Split Button plugin. Also added option to set initial style for button when creating it, and create button without icon.
```
Comment (Comandeer) [inline CHANGES.md]: Please add link to the plugin.

File: plugins/splitbutton/plugin.js (hunk 4 above)
Comment (Comandeer) [reply plugins/splitbutton/plugin.js]: Then I'd think about making `hide` method on button or any other toolbar element.

File: plugins/splitbutton/plugin.js [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+			item.role = 'menuitemcheckbox';
+
+			// We need both 'click' and 'onClick', because this definition can be used for setting both buttons and menu items.
+			// Button uses 'click' property, menu item uses 'onClick' property.
+			item.onClick = item.onClick || item.click ||  function() {
+				editor.execCommand( this.command, this.commandData );
+			};
+			item.click = item.onClick;
+
+			editor.getCommand( item.command ).on( 'state', function() {
```
Comment (Comandeer) [inline plugins/splitbutton/plugin.js]: It's still not possible to bind only click to button, without any command.

Comment (Comandeer) [reply plugins/splitbutton/plugin.js]: There aren't also tests for such case.

File: plugins/menu/plugin.js [hunk 7]
```diff
@@ -188,7 +188,7 @@ CKEDITOR.plugins.add( 'menu', {
 							var item = this.editor.getMenuItem( itemName );
 
 							if ( item && ( !item.command || this.editor.getCommand( item.command ).state ) ) {
-								item.state = listenerItems[ itemName ];
+								item.state = ( item.stateFn && item.stateFn() ) || listenerItems[ itemName ];
```
Comment (Comandeer) [inline plugins/menu/plugin.js]: I'm not sure if a better solution won't be to just substitute listener, not add another property directly to `menu`.

**Comment by engineering-this** [comment]
>I'm wondering also about one thing: why do we need to render and hide all buttons? Can't we have one button that would act as face and its icon and click actions would be just replaced ad hoc? Is the current solution connected with our convoluted way of registering commands alongside the buttons? If yes, then I really think if #678 isn't blocking.

@Comandeer: We need to render buttons, otherwise we won't register commands. And if thats necessary then I'm reusing all rendered buttons. It might be possible to remove them after rendering and just switch face onClick, icon and label, but that would just make all the logic even more complex. Imho this would be too haxy.

**Comment by engineering-this** [comment]
Rebased onto the latest major and updated version references to 4.11.

**Comment by engineering-this** [comment]
Rebased onto latest `major`.

**Pull request closed without merging**